nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/RetryFlowFile.java

```diff
@@ -0,0 +1,225 @@
+import org.apache.nifi.processor.Relationship;
+import org.apache.nifi.processor.exception.ProcessException;
+import org.apache.nifi.processor.util.StandardValidators;
+import org.apache.nifi.util.StringUtils;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+@Tags({"Retry", "FlowFile"})
+@CapabilityDescription("FlowFiles passed to this Processor have a 'Retry Attribute' value checked against a " +
+        "configured 'Maximum Retries' value. If the current attribute value is below the configured maximum, the " +
+        "FlowFile is passed to a retry relationship. The FlowFile may or may not be penalized in that condition. " +
+        "If the FlowFile's attribute value exceeds the configured maximum, the FlowFile will be passed to a " +
+        "'retries_exceeded' relationship. WARNING: If the incoming FlowFile has a non-numeric value in the " +
+        "configured 'Retry Attribute' attribute, it will be reset to '1'. You may choose to fail the FlowFile " +
+        "instead of performing the reset. Additional dynamic properties can be defined for any attributes you " +
+        "wish to add to the FlowFiles transferred to 'retries_exceeded'. These attributes support attribute " +
+        "expression language.")
+@InputRequirement(InputRequirement.Requirement.INPUT_REQUIRED)
+@SupportsBatching
+@SideEffectFree
+@DefaultSettings(penaltyDuration = "2 min")
+@ReadsAttribute(attribute = "Retry Attribute",
+        description = "Will read the attribute or attribute expression language result as defined in 'Retry Attribute'")
+@WritesAttribute(attribute = "Retry Attribute",
+        description = "User defined retry attribute is updated with the current retry count")
+@DynamicProperty(name = "Exceeded FlowFile Attribute Key",
+        value = "The value of the attribute added to the FlowFile",
+        description = "One or more dynamic properties can be used to add attributes to FlowFiles passed to " +
+                "the 'retries_exceeded' relationship",
+        expressionLanguageScope = ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
+public class RetryFlowFile extends AbstractProcessor {
+    private List<PropertyDescriptor> properties;
+    private Set<Relationship> relationships;
+    private String retryAttribute;
+    private Integer maximumRetries;
+    private Boolean penalizeRetried;
+    private Boolean failOnOverwrite;
+
+    public static final PropertyDescriptor RETRY_ATTRIBUTE = new PropertyDescriptor.Builder()
+            .name("Retry Attribute")
+            .description("The name of the attribute that contains the current retry count for the FlowFile. " +
+                    "WARNING: If the name matches an attribute already on the FlowFile that does not contain a " +
+                    "numerical value, the processor will either overwrite that attribute with '1' or fail " +
+                    "based on configuration.")
```

Comment:
Just an idea. I imagine users may experience unexpected behavior when they use RetryFlowFile processor multiple times for a FlowFile until it gets its end of life, in case if such RetryFlowFile processors are configured to use the same `Retry Attribute` at different places, the later RetryFlowFile retries less than expected, because retry count will be accumulated.

To avoid that, (and easier UX from being able to rely on the default attribute name), I'd suggest adding another property such as `Reset Previous Retry Attribute`, defaults to `true`. Add some code to write another FlowFile attribute tracking current RetryFlowFile instance identifier (via `getIdentifier()`). The attribute name can be static, e.g. `retryflowfile.processor.id`. Then at the beginning of `onTrigger()`, check if current identifier is the same with the one stored at the attribute. If different, reset `Retry Attribute` to 1.

The only possible use-case I can imagine where user would need disable doing this, is when RetryFlowFiles are nested. In such a case, each instance must be configured with different `Retry Attribute`.

How do you think?

Reply:
This was certainly something that I considered after opening the PR. I completely forgot about the `getIdentifier()` method on Processor. I'll implement this and the other changes when I have a moment.

Reply:
Please let me know when it gets ready for the next review cycle, thank you!

Reply:
I have updated the PR for the new behavior. It is now a drop-down selection for 1) Fail on Reuse 2) Warn on Reuse 3) Reset on Reuse. Let me know if you have any questions.